pkg/imageprovider/rhcos.go

```diff
@@ -84,8 +93,38 @@ func imageKey(data imageprovider.ImageData) string {
 	)
 }
 
+func getIgnitionOverride(imageMetadata *metav1.ObjectMeta, log logr.Logger) ([]byte, error) {
+	if overrideURI, exist := imageMetadata.Annotations[ignitionOverrideAnnotation]; exist {
+		log.Info("using Ignition override when building the image", "host", imageMetadata.Name, "overrideURI", overrideURI)
+		resp, err := http.Get(overrideURI) //#nosec G107
+		if err != nil {
+			return nil, errors.Wrap(err, "could not download Ignition override")
+		}
+		defer resp.Body.Close()
+
+		override, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not download Ignition override")
+		}
+
+		return override, nil
+	}
+
+	if infraEnvName, useInfraEnv := imageMetadata.Labels[infraEnvLabel]; useInfraEnv {
+		log.Info("host is using an InfraEnv, waiting for Ignition override", "host", imageMetadata.Name, "infraEnv", infraEnvName)
+		return nil, imageprovider.ImageNotReady{}
+	}
```

Comment:
Thinking out loud and keeping in mind the decoupling we have discussed. I wonder if checking for the InfraEnv label (and waiting for the ignition when it's present) is OK. It adds some layering and dependency as adding the label will prevent the metal3 workflow.

At first thought, I don't think it's an issue.

Reply:
At the very least, I'd start like this and iterate. Otherwise we'll never get it done :)